src/graph.rs

```diff
@@ -0,0 +1,343 @@
+        let index = SyntaxNodeRef(node.id() as SyntaxNodeID);
+        self.syntax_nodes.insert(index, node);
+        index
+    }
+
+    /// Adds a new graph node to the graph, returning a graph DSL reference to it.
+    pub fn add_graph_node(&mut self) -> GraphNodeRef {
+        let graph_node = GraphNode::new();
+        let index = self.graph_nodes.len() as GraphNodeID;
+        self.graph_nodes.push(graph_node);
+        GraphNodeRef(index)
+    }
+
+    /// Displays the contents of this graph.
+    pub fn display_with<'a>(&'a self, ctx: &'a Context) -> impl Display + 'a {
+        struct DisplayGraph<'a, 'tree>(&'a Graph<'tree>, &'a Context);
+
+        impl<'a, 'tree> Display for DisplayGraph<'a, 'tree> {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                let graph = self.0;
+                let ctx = self.1;
+                for (node_index, node) in graph.graph_nodes.iter().enumerate() {
+                    write!(
+                        f,
+                        "node {}\n{}",
+                        node_index,
+                        node.attributes.display_with(ctx, graph)
+                    )?;
+                    for (sink, edge) in &node.outgoing_edges {
+                        write!(
+                            f,
+                            "edge {} -> {}\n{}",
+                            node_index,
+                            *sink,
+                            edge.attributes.display_with(ctx, graph)
+                        )?;
+                    }
+                }
+                Ok(())
+            }
+        }
+
+        DisplayGraph(self, ctx)
+    }
+}
+
+impl<'tree> Index<SyntaxNodeRef> for Graph<'tree> {
+    type Output = Node<'tree>;
+    fn index(&self, index: SyntaxNodeRef) -> &Node<'tree> {
+        &self.syntax_nodes[&index]
+    }
+}
+
+impl Index<GraphNodeRef> for Graph<'_> {
+    type Output = GraphNode;
+    fn index(&self, index: GraphNodeRef) -> &GraphNode {
+        &self.graph_nodes[index.0 as usize]
+    }
+}
+
+impl<'tree> IndexMut<GraphNodeRef> for Graph<'_> {
+    fn index_mut(&mut self, index: GraphNodeRef) -> &mut GraphNode {
+        &mut self.graph_nodes[index.0 as usize]
+    }
+}
+
+/// A node in a graph
+pub struct GraphNode {
+    outgoing_edges: SmallVec<[(GraphNodeID, Edge); 8]>,
+    /// The set of attributes associated with this graph node
+    pub attributes: Attributes,
+}
+
+impl GraphNode {
+    fn new() -> GraphNode {
+        GraphNode {
+            outgoing_edges: SmallVec::new(),
+            attributes: Attributes::new(),
+        }
+    }
+
+    /// Adds an edge to this node.  There can be at most one edge connecting any two graph nodes;
+    /// the result indicates whether the edge is new (`Ok`) or already existed (`Err`).  In either
+    /// case, you also get a mutable reference to the [`Edge`][] instance for the edge.
+    pub fn add_edge(&mut self, sink: GraphNodeRef) -> Result<&mut Edge, &mut Edge> {
```

Comment:
Instead of failing when the edge already exists, could we union the attributes of the extant and new `Edge` and return successfully?

Reply:
This has come up a couple of times now.  If we went with a more Datalog-inspired DSL, where the execution order of stanzas isn't important, then this kind of unioning of content would be an important part of that.  But as written so far, the language is purposefully more imperative in nature, and so I wanted to minimize the amount of work that's done transparently for you.  I'd suggest that in that version of the language, unioning a set of attributes should be a function that we provide that you can call yourself when (and only when) you want that behavior.